cssparse/shared/src/main/scala/cssparse/CssParser.scala

```diff
@@ -0,0 +1,205 @@
+
+  // any token except function_token
+  val simple_token: Parser[Option[Ast.SimpleToken]] = P(
+    whitespace_token    | at_word_token |
+    hash_token          | match_token |
+    column_token        | CDO_token |
+    CDC_token           | string_token |
+    unicode_range_token | percentage_token |
+    dimension_token     | url_token |
+    number_token        | ident_token |
+    delim_token ).map({
+      case st: Ast.SimpleToken => Some(st)
+      case _ => None
+    })
+
+  val bracketsBlock =       P( "(" ~ componentValue.rep ~ ")" ).map(values => Ast.BracketsBlock(values.flatten))
+  val curlyBracketsBlock =  P( "{" ~ componentValue.rep ~ "}" ).map(values => Ast.CurlyBracketsBlock(values.flatten))
+  val squareBracketsBlock = P( "[" ~ componentValue.rep ~ "]" ).map(values => Ast.SquareBracketsBlock(values.flatten))
+
+  val functionBlock = P( function_token ~ componentValue.rep ~ ")").map({
+    case (Ast.FunctionToken(name), values: Seq[Option[Ast.ComponentValue]]) =>
+      Ast.FunctionBlock(name, Ast.BracketsBlock(values.flatten))
+  })
+
+  val componentValue: Parser[Option[Ast.ComponentValue]] = P( simple_token | bracketsBlock |
+    curlyBracketsBlock | squareBracketsBlock | functionBlock ).map({
+      case cv: Ast.ComponentValue => Some(cv)
+      case Some(cv: Ast.ComponentValue) => Some(cv)
+      case _ => None
+    })
+}
+
+object CssRulesParser {
+
+  import CssTokensParser._
+
+  val allSelector = P( "*" ).map({_ => Ast.AllSelector()})
+
+  val elementSelector = P( ident_token.! ).map(Ast.ElementSelector)
+
+  val idSelector = P( "#" ~ ident_token.! ).map(Ast.IdSelector)
+
+  val attributeSelector = P( ident_token.!.? ~ ("[" ~ ident_token.! ~
+    (( "=" | match_token).! ~ (string_token | ident_token)).? ~ "]").rep(1) ).map({
+    case (name, attrs) => Ast.AttributeSelector(name, attrs.map({
+      case (ident, Some((token, Ast.StringToken(string)))) => (ident, Some(token), Some(string))
+      case (ident, Some((token, Ast.IdentToken(string)))) => (ident, Some(token), Some(string))
+      case (ident, None) => (ident, None, None)
+    }))
+  })
+
+  val partSelector = P( allSelector | attributeSelector | elementSelector )
+
+  val classSelectorPart = P( "." ~  partSelector ).map(Ast.ClassSelectorPart)
+
+  val pseudoSelectorPart = P( (("::" | ":") ~ ident_token).! ~ ("(" ~ componentValue.rep(1) ~ ")").? ).map({
+    case (name, optValues) =>
+      Ast.PseudoSelectorPart(name, optValues.map(cvs => cvs.flatMap(cv => cv)).getOrElse(ArrayBuffer()))
+  })
+
+  val complexSelectorPart = P( pseudoSelectorPart | classSelectorPart )
+
+  val complexSelector = P( partSelector.? ~ complexSelectorPart.rep(1) ).map({
+    case (part, parts) => Ast.ComplexSelector(part, parts)
+  })
+
+  val singleSelector: Parser[Ast.SingleSelector] = P( complexSelector | partSelector | idSelector | allSelector )
+
+  val selectorDelim = P( (ws ~ CharIn(",>+~").! ~ ws) | whitespace_token.rep(1).! ).map({
+    case s if s.startsWith(" ") => " "
```

Comment:
Because it's a selector delimiter, and if wrote `ws ~ whitespace_token.rep(1) ~ ws` or just `ws` it wouldn't work since `ws` can mean any number of spaces including 0.